From 5475e17858d143747e69b1bf9e8d230e74642561 Mon Sep 17 00:00:00 2001
From: David Vossel <dvossel@redhat.com>
Date: Fri, 11 Jul 2014 11:22:20 -0400
Subject: [PATCH] Low: nfsnotify: fixes error message output

---
 heartbeat/nfsnotify | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/heartbeat/nfsnotify b/heartbeat/nfsnotify
index 2e242de..2d0bbfc 100755
--- a/heartbeat/nfsnotify
+++ b/heartbeat/nfsnotify
@@ -269,7 +269,7 @@ v3notify_start()
 		ocf_log info "sending notifications with source address $ip"
 		$SM_NOTIFY_BINARY -f $OCF_RESKEY_notify_args -v $ip -P "$cur_statd"
 		if [ $? -ne 0 ]; then
-			ocf_log err "sm-notify with source host set to, $source_host, failed. view syslog for more information"
+			ocf_log err "sm-notify with source host set to, $ip, failed. view syslog for more information"
 			return $OCF_ERR_GENERIC
 		fi
 	done
-- 
1.8.4.2

