From e5c5c087ecf152bd69f5795024bfc655394c3c18 Mon Sep 17 00:00:00 2001
From: Andreas Kurz <andreas.kurz@gmail.com>
Date: Thu, 18 Sep 2014 23:06:36 +0200
Subject: [PATCH 1/6] High: galera: do not ignore specified check_password

---
 heartbeat/galera | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/heartbeat/galera b/heartbeat/galera
index 54654f8..386daaf 100755
--- a/heartbeat/galera
+++ b/heartbeat/galera
@@ -672,7 +672,7 @@ fi
 
 MYSQL_OPTIONS_CHECK="-nNE --user=${OCF_RESKEY_check_user}"
 if [ -n "${OCF_RESKEY_check_passwd}" ]; then
-    MYSQL_OPTIONS_CHECK="$MYSQL_OPTIONS_CHECK --password=${MYSQL_PASSWORD}"
+    MYSQL_OPTIONS_CHECK="$MYSQL_OPTIONS_CHECK --password=${OCF_RESKEY_check_passwd}"
 fi
 
 # What kind of method was invoked?
-- 
1.8.4.2

