From bc1e7bdcedc1bb1bf473787f373261452e37e337 Mon Sep 17 00:00:00 2001
From: David Vossel <dvossel@redhat.com>
Date: Wed, 30 Jul 2014 12:59:46 -0500
Subject: [PATCH 5/6] Low: galera: do not advertise notify in the usage

---
 heartbeat/galera | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/heartbeat/galera b/heartbeat/galera
index 386daaf..a361d7b 100755
--- a/heartbeat/galera
+++ b/heartbeat/galera
@@ -79,7 +79,7 @@ fi
 
 usage() {
   cat <<UEND
-usage: $0 (start|stop|validate-all|meta-data|monitor|promote|demote|notify)
+usage: $0 (start|stop|validate-all|meta-data|monitor|promote|demote)
 
 $0 manages a galera Database as an HA resource.
 
@@ -237,7 +237,6 @@ Cluster check user password
 <action name="monitor" role="Slave" depth="0" timeout="30" interval="30" />
 <action name="promote" timeout="120" />
 <action name="demote" timeout="120" />
-<action name="notify" timeout="90" />
 <action name="validate-all" timeout="5" />
 <action name="meta-data" timeout="5" />
 </actions>
@@ -683,7 +682,6 @@ case "$1" in
   monitor)  galera_monitor;;
   promote)  galera_promote;;
   demote)   galera_demote;;
-  notify)   galera_notify;;
   validate-all) exit $OCF_SUCCESS;;
 
  *)     usage
-- 
1.8.4.2

