diff -uNr a/heartbeat/NovaEvacuate b/heartbeat/NovaEvacuate
--- a/heartbeat/NovaEvacuate	2016-02-29 10:54:21.933786269 +0100
+++ b/heartbeat/NovaEvacuate	2016-02-29 13:29:27.000139496 +0100
@@ -177,17 +177,10 @@
 	esac
 
 	if [ $need_evacuate = 1 ]; then
-	    found=0
 	    ocf_log notice "Initiating evacuation of $node"
 
-	    for known in $(fence_compute ${fence_options} -o list | tr -d ','); do
-		if [ ${known} = ${node} ]; then
-		    found=1
-		    break
-		fi
-	    done
-
-	    if [ $found = 0 ]; then
+	    fence_compute ${fence_options} -o status -n ${node}
+	    if [ $? != 0 ]; then
 		ocf_log info "Nova does not know about ${node}"
 		# Dont mark as no because perhaps nova is unavailable right now
 		continue
