From cbccff5ed9b1fc5641063f05ad531f897d366fa4 Mon Sep 17 00:00:00 2001
From: Mike Bayer <mike_mp@zzzcomputing.com>
Date: Tue, 15 Sep 2015 14:54:05 -0400
Subject: [PATCH] galera: add support for MYSQL_HOST and MYSQL_PORT from
 clustercheck

---
 heartbeat/galera | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/heartbeat/galera b/heartbeat/galera
index 920507b..1a1a4ce 100755
--- a/heartbeat/galera
+++ b/heartbeat/galera
@@ -704,6 +704,18 @@ if [ -n "${OCF_RESKEY_check_passwd}" ]; then
     MYSQL_OPTIONS_CHECK="$MYSQL_OPTIONS_CHECK --password=${OCF_RESKEY_check_passwd}"
 fi
 
+# This value is automatically sourced from /etc/sysconfig/checkcluster if available
+if [ -n "${MYSQL_HOST}" ]; then
+    MYSQL_OPTIONS_CHECK="$MYSQL_OPTIONS_CHECK -h ${MYSQL_HOST}"
+fi
+
+# This value is automatically sourced from /etc/sysconfig/checkcluster if available
+if [ -n "${MYSQL_PORT}" ]; then
+    MYSQL_OPTIONS_CHECK="$MYSQL_OPTIONS_CHECK -P ${MYSQL_PORT}"
+fi
+
+
+
 # What kind of method was invoked?
 case "$1" in
   start)    galera_start;;
