diff -uNr a/heartbeat/Filesystem b/heartbeat/Filesystem
--- a/heartbeat/Filesystem	2016-12-20 14:16:29.439700386 +0100
+++ b/heartbeat/Filesystem	2016-12-20 14:18:04.954623779 +0100
@@ -538,6 +538,12 @@
 		Filesystem_notify
 	fi
 
+	# Check if there are any mounts mounted under the mountpoint
+	if list_mounts | grep -q -E " $MOUNTPOINT/\w+" >/dev/null 2>&1; then
+		ocf_log err "There is one or more mounts mounted under $MOUNTPOINT."
+		return $OCF_ERR_CONFIGURED
+	fi
+
 	# See if the device is already mounted.
 	if Filesystem_status >/dev/null 2>&1 ; then
 		ocf_log info "Filesystem $MOUNTPOINT is already mounted."
