From 051743955c4f1f5fe412875afba94edd2839008c Mon Sep 17 00:00:00 2001
From: Oyvind Albrigtsen <oalbrigt@redhat.com>
Date: Wed, 22 Nov 2017 12:25:41 +0100
Subject: [PATCH] db2: fix HADR promote when master failed

---
 heartbeat/db2 | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/heartbeat/db2 b/heartbeat/db2
index 63de31582..b67363ec5 100755
--- a/heartbeat/db2
+++ b/heartbeat/db2
@@ -617,7 +617,7 @@ db2_instance_status() {
     if [ $pscount -ge 4 ]; then
         return $OCF_SUCCESS;
     elif [ $pscount -ge 1 ]; then
-        return $OCF_GENERIC_ERR
+        return $OCF_ERR_GENERIC
     fi
     return $OCF_NOT_RUNNING
 }
@@ -767,7 +767,7 @@ db2_promote() {
             # must take over 
             ;;
 
-            STANDBY/PEER/DISCONNECTED|Standby/DisconnectedPeer)
+            STANDBY/PEER/DISCONNECTED|STANDBY/DISCONNECTED_PEER/DISCONNECTED|Standby/DisconnectedPeer)
             # must take over forced 
             force="by force peer window only"
             ;;
