diff --git a/heartbeat/Makefile.am b/heartbeat/Makefile.am
index 3a05974..777c6b1 100644
--- a/heartbeat/Makefile.am
+++ b/heartbeat/Makefile.am
@@ -25,7 +25,7 @@ EXTRA_DIST		= $(ocf_SCRIPTS) $(ocfcommon_DATA) \
 
 INCLUDES		= -I$(top_srcdir)/include -I$(top_srcdir)/linux-ha
 
-halibdir		= $(libdir)/heartbeat
+halibdir		= $(libexecdir)/heartbeat
 
 ocfdir		        = $(OCF_RA_DIR_PREFIX)/heartbeat
 
diff --git a/heartbeat/ocf-directories.in b/heartbeat/ocf-directories.in
index 26004a3..6e0a9d5 100644
--- a/heartbeat/ocf-directories.in
+++ b/heartbeat/ocf-directories.in
@@ -11,7 +11,7 @@ exec_prefix=@exec_prefix@
 : ${HA_RSCTMP:=@HA_RSCTMPDIR@}
 : ${HA_RSCTMP_OLD:=@HA_VARRUNDIR@/heartbeat/rsctmp}
 : ${HA_FIFO:=@localstatedir@/lib/heartbeat/fifo}
-: ${HA_BIN:=@libdir@/heartbeat}
+: ${HA_BIN:=@libexecdir@/heartbeat}
 : ${HA_SBIN_DIR:=@sbindir@}
 : ${HA_DATEFMT:="%Y/%m/%d_%T "}
 : ${HA_DEBUGLOG:=/dev/null}
diff --git a/resource-agents.spec.in b/resource-agents.spec.in
index c306398..e9f0688 100644
--- a/resource-agents.spec.in
+++ b/resource-agents.spec.in
@@ -255,7 +255,7 @@ rm -rf %{buildroot}
 %dir %{_sysconfdir}/ha.d
 %{_sysconfdir}/ha.d/shellfuncs
 
-%{_libdir}/heartbeat
+%{_libexecdir}/heartbeat
 
 %post -n resource-agents
 if [ $1 = 2 ]; then
diff --git a/tools/Makefile.am b/tools/Makefile.am
index 971db1c..3205bbd 100644
--- a/tools/Makefile.am
+++ b/tools/Makefile.am
@@ -23,7 +23,7 @@ SUBDIRS			= ocft
 
 INCLUDES                = -I$(top_builddir)/include -I$(top_srcdir)/include
 
-halibdir		= $(libdir)/heartbeat
+halibdir		= $(libexecdir)/heartbeat
 
 EXTRA_DIST		= ocf-tester.8 sfex_init.8
 

